utils/land-patterns/ipc-helpers.stanza

```diff
@@ -0,0 +1,314 @@
+      LevelC: LeadFillets(0.3, -0.20, 0.3, 0.25)
+  else:
+    switch(p):
+      LevelA: LeadFillets(1.0,  0.00, 0.6, 1.0)
+      LevelB: LeadFillets(0.7, -0.05, 0.5, 0.5)
+      LevelC: LeadFillets(0.4, -0.10, 0.4, 0.25)
+
+protected defn aluminum-electrolytic (p:DesignProducability, height:Double):
+  two-pin-crystal(p, height)
+
+;==============================================================================
+;========================== Main Helpers ======================================
+;==============================================================================
+defpackage ocdb/land-patterns/ipc-helpers:
+  import core
+  import collections
+  import math
+  import jitx
+  import jitx/commands
+  import ocdb/design-vars
+  import ocdb/land-patterns/leads
+  import ocdb/land-patterns/packages
+
+; Component tolerances, defined by component manufacturers.
+public pcb-struct ocdb/land-patterns/ipc-helpers/ComponentTolerance:
+  width: Double,      ; tolerance on the width of a land pattern
+  length:Double,      ; tolerance on the length of a land pattern
+  lead-widths:Double, ; tolerance on the terminations or leads of a land pattern
+
+; A single value and its tolerance
+public pcb-struct ocdb/land-patterns/ipc-helpers/SingleTolerance:
+  value:Double, 
+  tolerance:Double, ; TODO with: ensure => non-negative! and less than value
+
+; A set of manufacturing tolerances
+public pcb-struct ocdb/land-patterns/ipc-helpers/ManufacturerTolerances:
+  fabrication:Double,       ; unilateral profile tolerance for the printed board land pattern 
+  placement-accuracy:Double ; diameter of true position placement accuracy to the center of the land pattern
+
+; A single value and its tolerance
+public pcb-struct ocdb/land-patterns/ipc-helpers/SingleTolerance:
+  nominal-value:Double, 
+  tolerance:Double, ; TODO with: ensure => non-negative! and less than value
+
+; Return the minimum value for a value with some tolerance
+public defn min (s:SingleTolerance) -> Double: 
+  value(s) - tolerance(s)
+
+; Return the maximum value for a value with some tolerance
+public defn max (s:SingleTolerance) -> Double: 
+  value(s) + tolerance(s)
+
+; Find the RMS sum of many tolerances. This operation is non-commutative and non-associative, so 
+; it should be called once when working with tolerances.
+public defn sum (s:Seqable<SingleTolerance>) -> SingleTolerance:
+  var nom = 0.0
+  var tol = 0.0
+  for t in s do:
+    nom = nom + value(t)
+    tol = tol + tolerance(s) * tolerance(s)
+  SingleTolerance(nom, sqrt(tol))
+
+;==============================================================================
+; Value References:
+;
+; Z: Double,  ; overall length of the landpattern
+; G: Double,  ; distance between pads
+; X: Double,  ; width of the landpattern
+; L: Double,  ; overall length of the component
+; S: Double,  ; distance between component terminations
+; W: Double,  ; width of the lead or termination
+; Jt: Double, ; solder fillet at toe
```

Comment:
where is it used?

Reply:
This is pulled from the IPC reference. I leave it in here in case datasheets use these variables and need to translate to internal data structures, as well as references for the internals of the helpers.